File: FlyingDanger/FlyingDangerC++/Castle.cpp
#include "Castle.hpp"

int currentHealth = 100;
int maxHealth = 100;

Castle::Castle() : health(currentHealth), max_health(maxHealth), position(""), sprite(""), rect("") {
    cout << "Объект класса Castle создан" << endl;
}

Castle::Castle(int max_hp) : health(max_hp), max_health(max_hp), position(""), sprite(""), rect("") {
    cout << "Объект класса Castle создан с максимальным здоровьем: " << max_hp << endl;
}

Castle::~Castle() {
    cout << "Объект класса Castle уничтожен" << endl;
}

void Castle::TakeDamage(int amount) {
    health -= amount;
    cout << "Метод take_damage() класса Castle выполнен. Получено урона: " << amount << endl;
    cout << "Осталось здоровья замка: " << health << "/" << max_health << endl;
}

bool Castle::IsDestroyed() {
    bool destroyed = health <= 0;
    cout << "Метод is_destroyed() класса Castle выполнен. Результат: " << (destroyed ? "разрушен" : "цел") << endl;
    return destroyed;
}

void Castle::Draw() {
    cout << "Метод draw() класса Castle выполнен" << endl;
}

File: YaroslavSychev/FlyingDangerJava/src/Castle.java
public class Castle {
    private int health;
    private int maxHealth;
    private String position;
    private String sprite;
    private String rect;

    // Константы
    public static final int DEFAULT_HEALTH = 100;
    public static final int DEFAULT_MAX_HEALTH = 100;
    public static final int CASTLE_ATTACK_DAMAGE = 10;

    public Castle() {
        this.health = DEFAULT_HEALTH;
        this.maxHealth = DEFAULT_MAX_HEALTH;
        this.position = "";
        this.sprite = "";
        this.rect = "";
        System.out.println("Объект класса Castle создан");
    }

    public Castle(int maxHealth) {
        this.health = maxHealth;
        this.maxHealth = maxHealth;
        this.position = "";
        this.sprite = "";
        this.rect = "";
        System.out.println("Объект класса Castle создан с максимальным здоровьем: " + maxHealth);
    }

    public void takeDamage(int amount) {
        if (amount < 0) {
            throw new IllegalArgumentException("Урон не может быть отрицательным: " + amount);
        }
        if (amount > 1000) {
            throw new IllegalStateException("Слишком большой урон: " + amount);
        }

        try {
            health -= amount;
            if (health < 0) {
                health = 0;
                throw new IllegalStateException("Замок разрушен!");
            }
            System.out.println("Метод takeDamage() класса Castle выполнен. Получено урона: " + amount);
            System.out.println("Осталось здоровья замка: " + health + "/" + maxHealth);
        } catch (IllegalStateException e) {
            System.out.println("КРИТИЧЕСКАЯ ОШИБКА: " + e.getMessage());
        }
    }

        public boolean isDestroyed() {
        boolean destroyed = health <= 0;
        System.out.println("Метод isDestroyed() класса Castle выполнен. Результат: " +
                (destroyed ? "разрушен" : "цел"));
        return destroyed;
    }

    public void draw() {
        System.out.println("Метод draw() класса Castle выполнен");
    }

    public int getHealth() { return health; }
    public int getMaxHealth() { return maxHealth; }
}